### PR TITLE
Addition of myDewControllerPro driver

### PR DESCRIPTION
**Pull request by chemistorge**
Addition of myDewController Driver for R Browns dew controller. https://sourceforge.net/projects/arduinonanodewcontrollerpro/files/myDewControllerPro%20v300%203channel/

### CODE REVIEW

**Comment by chemistorge** [comment]
Hi, 

Once this driver gets added to indiLib, how does Kstars/ekos get updated? For testing i had to manual add an XML file under /usr/share/indi for ekos to see the new driver. 

Sorry if this is a silly question but I am new to this contributing and github.

**Comment by knro** [comment]
You can sync your fork or simply rebase to master since your includes actually revert INDI commits. or you can close this PR and re-submit

File: drivers/auxiliary/myDewControllerPro.h [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        bool setFanTempTrigger(int tempOn, int tempOff);
+        bool zeroTempCalibrations();
+        bool setInt(int mode, const char* mask, const char* errMessage);
+        bool setChoice(int testInt, const char* positiveChoice, const char* negativeChoice, const char* errMessage);
+
+        Connection::Serial *serialConnection{ nullptr };
+        int PortFD{ -1 };
+
+        INumber OutputsN[3];
+        INumberVectorProperty OutputsNP;
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.h:154]: Please use modern INDI properties, i.e.

INDI::PropertyNumber OutputsNP {3}

File: drivers/auxiliary/myDewControllerPro.h [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+        enum {
+            DEW_STRAP_ONE_POWER,
+            DEW_STRAP_TWO_POWER,
+            DEW_STRAP_THREE_POWER
+        };
+
+        ISwitch CH1CH2BoostS[2];
+        ISwitchVectorProperty CH1CH2BoostSP;
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.h:163]: INDI::PropertySwitch CH1CH2BoostSP {2}

File: drivers/auxiliary/myDewControllerPro.h [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        INumberVectorProperty CH3_Manual_PowerNP;
+
+        INumber FanSpeedN[1];
+        INumberVectorProperty FanSpeedNP;
+
+        ISwitch FanModeS[2];
+        ISwitchVectorProperty FanModeSP;
+
+        INumber TemperaturesN[5];
+        INumberVectorProperty TemperaturesNP;
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.h:191]: Use modern INDI::PropertyXXX properties, not the legacy properties.

File: drivers/auxiliary/myDewControllerPro.h [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+
+        ISwitch EnableLCDDisplayS[2];
+        ISwitchVectorProperty EnableLCDDisplaySP;
+
+        INumber LCDPageRefreshN[1];
+        INumberVectorProperty LCDPageRefreshNP;
+
+        ISwitch EEPROMS[2];
+        ISwitchVectorProperty EEPROMSP;
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.h:264]: Use modern properties

File: drivers/auxiliary/myDewControllerPro.cpp [hunk 5]
```diff
@@ -0,0 +1,945 @@
...
+
+bool myDewControllerPro::initProperties()
+{
+    DefaultDevice::initProperties();
+
+    /* Channel duty cycles */
+    IUFillNumber(&OutputsN[DEW_STRAP_ONE_POWER], "CHANNEL1", "Strap 1", "%4.0f %%", 0., 100., 1., 0.);
+    IUFillNumber(&OutputsN[DEW_STRAP_TWO_POWER], "CHANNEL2", "Strap 2", "%4.0f %%", 0., 100., 1., 0.);
+    IUFillNumber(&OutputsN[DEW_STRAP_THREE_POWER], "CHANNEL3", "Strap 3", "%4.0f %%", 0., 100., 1., 0.);
+    IUFillNumberVector(&OutputsNP, OutputsN, 3, getDeviceName(), "OUTPUT", "Outputs", MAIN_CONTROL_TAB, IP_RO, 0, IPS_IDLE);
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.cpp:48]: OutputsNP[DEW_STRAP_ONE_POWER].fill("CHANNEL1", "Strap 1", "%4.0f %%", 0., 100., 1., 0.);

...etc.

File: drivers/auxiliary/myDewControllerPro.cpp [hunk 6]
```diff
@@ -0,0 +1,945 @@
...
+        }
+        if (!readOffsetValues()) {
+            LOG_INFO("Reading Offset Values Error");
+        }
+        LOG_INFO("myDewControllerPro parameters updated, device ready for use.");
+        SetTimer(getCurrentPollingPeriod());
+    }
+    else
+    {
+        deleteProperty(OutputsNP.name);
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.cpp:215]: deleteProperty(OutputsNP);

File: drivers/auxiliary/myDewControllerPro.cpp [hunk 7]
```diff
@@ -0,0 +1,945 @@
...
+
+}
+
+bool myDewControllerPro::updateProperties()
+{
+    DefaultDevice::updateProperties();
+
+    if (isConnected())
+    {
+        defineProperty(&OutputsNP);
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.cpp:174]: defineProperty(OutputsNP);

File: drivers/auxiliary/myDewControllerPro.cpp [hunk 8]
```diff
@@ -0,0 +1,945 @@
...
+        if (CH1CH2BoostS[CH1_BOOST_100].s == ISS_ON) {
+            setOutputBoost(CH1_BOOST_100);
+        }
+        if (CH1CH2BoostS[CH2_BOOST_100].s == ISS_ON) {
+            setOutputBoost(CH2_BOOST_100);
+        }
+        CH1CH2BoostSP.s = IPS_OK;
+        IDSetSwitch(&CH1CH2BoostSP, nullptr);
+        readMainValues();
+        return true;
```
Comment (knro) [inline drivers/auxiliary/myDewControllerPro.cpp:481]: Example below on how to change this to modern INDI convention. Please apply it to the rest of the code.

CH1CH2BoostSP.update(states, names, n);
CH1CH2BoostSP.setState(IPS_BUSY);
....
if (CH1CH2BoostSP[CH1_BOOST_100].getState() == ISS_ON) {
.....
CH1CH2BoostSP.setState(IPS_OK);
CH1CH2BoostSP.apply();

**Review by knro — CHANGES_REQUESTED** [review]
Please update to modern INDI convention. Add XML entry to drivers.xml file as well.

File: drivers/auxiliary/myDewControllerPro.h (hunk 1 above)
Comment (chemistorge) [reply drivers/auxiliary/myDewControllerPro.h:154]: I started using that but thought it was old not new as most drivers are the way I wrote. Where is drivers.xml located please.

File: drivers/auxiliary/myDewControllerPro.cpp (hunk 8 above)
Comment (chemistorge) [reply drivers/auxiliary/myDewControllerPro.cpp:481]: No problem, I know how to do that as I started that way and changed all my code (opps) as I thought it was wrong.

**Comment by chemistorge** [comment]
All changes made.

**Comment by chemistorge** [comment]
Thank you,

How long before the driver will be included in Ubuntu packages from the
PPA?. When indiLib is available as package with new driver will the driver
be seen by kstars/ekos?

Regards

Mark

**Comment by knro** [comment]
It should be in the nightly PPA and should be seen from any client parsing the drivers.xml file. Stable INDI 2.0.1 release is on March 31st.

**Comment by chemistorge** [comment]
Thank you

**Pull request merged**